src/gbm/gbtree.h

```diff
@@ -246,6 +246,14 @@ class GBTree : public GradientBooster {
   std::unique_ptr<Predictor> const& GetPredictor(HostDeviceVector<float> const* out_pred = nullptr,
                                                  DMatrix* f_dmat = nullptr) const {
     CHECK(configured_);
+    auto on_device = f_dmat && (*(f_dmat->GetBatches<SparsePage>().begin())).data.DeviceCanRead();
```

Comment:
This is a nasty expression, consider rewriting this so it is readable.

Reply:
Will implement a better method cooperating external memory version of ellpack page.  Right now for this PR I can't do that much of refactoring.  As going through all dmatrix implementations with batch iterator is not trivial.